cli/deno_dir.rs

```diff
@@ -69,16 +146,15 @@ pub struct DenoDir {
   // This is where we cache http resources. Example:
   // /Users/rld/.deno/deps/github.com/ry/blah.js
   pub gen: PathBuf,
+  pub gen_cache: DiskCache,
   // In the Go code this was called CacheDir.
   // This is where we cache compilation outputs. Example:
   // /Users/rld/.deno/gen/f39a473452321cacd7c346a870efb0e3e1264b43.js
   pub deps: PathBuf,
+  pub deps_cache: DiskCache,
```

Comment:
I was thinking the DiskCache would encompass both the "gen" and "deps" directories, not one for each.

Also isn't the `deps: PathBuf` superfluous, since one could access `deps_cache.location` to get the same value?

Reply:
> Also isn't the `deps: PathBuf` superfluous, since one could access `deps_cache.location` to get the same value?

Yes

> I was thinking the DiskCache would encompass both the "gen" and "deps" directories, not one for each.

I'm not sure... They you'd have to be aware what directory you want to generate filename for. I was thinking that `DenoDir` would encompass both `DiskCaches` storing them under common location (`$DENO_DIR`). If we expose public API for compilers then each compiler could create it's own cache.

Reply:
Hrm. Ok. 

I guess I'm saying this because I'm not entirely convinced that the gen/deps split is useful. We've talked in the past about getting rid of the gen directory and storing generated files next to their ts counterparts. But that's not going to happen in this change anyway, so I retract my comment.

Reply:
I do believe it's useful - this is best visible when you think about JS->JS compilation for remote module:
```
// suppose you import this module
import "https://github.com/foo/bar.js

// it will be downloaded and cached in:
$DENO_DIR/deps/https/github.com/foo/bar.js

// with `checkJs` enabled we're gonna have JS->JS compilation, and compiled file will be stored in:
$DENO_DIR/gen/https/github.com/foo/bar.js

// if we did not use `gen/` directory then compiled file would override downloaded file
```

Reply:
+1 for `deps/` + `gen/` with similar directory hierarchy. I still believe putting generated files next to source is not the best idea as it would make certain behavior "magically not working".

Haven't reviewed the whole change but I even prefer to move `.headers.json` files into another `meta/` directory.

Reply:
> +1 for `deps/` + `gen/` with similar directory hierarchy. I still believe putting generated files next to source is not the best idea as it would make certain behavior "magically not working".
> 
> Haven't reviewed the whole change but I even prefer to move `.headers.json` files into another `meta/` directory.

I really don't see benefit of moving `.headers.json` to separate directory. `deps/` is cache for downloaded modules and their metadata live next to them. The same applies for `.meta` in `gen/` directory.